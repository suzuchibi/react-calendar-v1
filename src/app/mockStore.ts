import configureMockStore from 'redux-mock-store';
const mock = configureMockStore();

export const mockStore = mock({
  drawer: false,
  today: 1626015600000,
  modal: {
    isActive: false,
    mode: '0',
  },
  form: {
    id: '0',
    isDuplicate: false,
  },
  cal: {
    current: 1625065200000,
    holidays: [
      {
        date: '2021-07-22',
        summary: '海の日',
      },
    ],
    status: 'loaded',
  },
  calPoper: {
    anchorEl: null,
  },
  schedule: {
    holidays: [],
    selected: 1626015600000,
    status: 'loaded',
  },
  datas: [
    {
      id: '1',
      title: 'タイトル1',
      coment: 'コメント1',
      start: 1626139200000,
      end: 1626141600000,
    },
  ],
  screen: false,
  snack: {
    open: false,
    msg: '',
  },
});
