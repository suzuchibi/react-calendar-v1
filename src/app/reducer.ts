import { combineReducers } from 'redux';
import isDrawerReducer from '../features/side/isDrawerReducer';
import todayReducer from '../features/today/todayReducer';
import modalReducer from '../features/modal/reducer';
import formReducer from '../features/form/reducer';
import calendarReducer from '../features/calendar/calendarReducer';
import calendarPoperReducer from '../features/dialog/calendarPoper/reducer';
import scheduleReducer from '../features/scheduler/reducers/scheduleReducer';
import scheduleDatasReducer from '../features/scheduler/reducers/datasReducer';
import fullScreenDialogReducer from '../features/dialog/fullScreen/reducer';
import snackReducer from '../features/snack/snackReducer';

const rootReducer = combineReducers({
  drawer: isDrawerReducer,
  today: todayReducer,
  modal: modalReducer,
  form: formReducer,
  cal: calendarReducer,
  calPoper: calendarPoperReducer,
  schedule: scheduleReducer,
  datas: scheduleDatasReducer,
  screen: fullScreenDialogReducer,
  snack: snackReducer,
});

export type RootState = ReturnType<typeof rootReducer>;
export default rootReducer;
