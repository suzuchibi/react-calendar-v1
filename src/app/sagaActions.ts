import { Holidays } from '../features/calendar/types';

const GOOGLE_CALENDAR_API = 'app/googleCalendarAPI';
const GOOGLE_CALENDAR_API_SUCESS = 'app/googleCalendarApiSuccess';
const GOOGLE_CALENDAR_API_ERROR = 'app/googleCalendarApiError';

export interface Props {
  mode: string;
  stamp: number;
}

export interface GoogleCalendarApiProps {
  type: typeof GOOGLE_CALENDAR_API;
  payload: Props;
}

interface HolidayProps {
  mode: string;
  result: Array<Holidays>;
}

export interface GoogleCalendarApiSuccessProps {
  type: typeof GOOGLE_CALENDAR_API_SUCESS;
  payload: HolidayProps;
}

export interface GoogleCalendarApiErrorProps {
  type: typeof GOOGLE_CALENDAR_API_ERROR;
  payload: {
    mode: string;
  };
}

export type ActionTypes =
  | GoogleCalendarApiProps
  | GoogleCalendarApiSuccessProps
  | GoogleCalendarApiErrorProps;

/**
 * google calendar API saga start
 * @param {object} payload
 */
export const gApiSaga = (payload: Props): ActionTypes => ({
  type: GOOGLE_CALENDAR_API,
  payload,
});

/**
 * Success
 * @param {object} payload
 */
export const gApiSuccess = (payload: HolidayProps): ActionTypes => ({
  type: GOOGLE_CALENDAR_API_SUCESS,
  payload,
});

/**
 * Error
 * @param {object} payload
 */
export const gApiError = (payload: { mode: string }): ActionTypes => ({
  type: GOOGLE_CALENDAR_API_ERROR,
  payload,
});
