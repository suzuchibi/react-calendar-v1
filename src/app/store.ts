import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import createSagaMiddleware from 'redux-saga';

import rootSaga from './sagas';
import rootReducer from './reducer';

const saga = createSagaMiddleware();
const store = createStore(
  rootReducer,
  composeWithDevTools(applyMiddleware(saga))
);
saga.run(rootSaga);

export default store;
