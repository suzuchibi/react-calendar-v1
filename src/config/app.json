{
  "drawerWidth": 260,
  "timeLineWidth": 60,
  "startOrder": 10,
  "lastOrder": 19,
  "dayOfTheWeek": ["日", "月", "火", "水", "木", "金", "土"],
  "gapiCal": {
    "key": "AIzaSyCLqi5tQwWw8qEGAWc9NC1cEU78C4EAWkM",
    "address": "ja.japanese#holiday@group.v.calendar.google.com",
    "gurl": "https://www.googleapis.com/calendar/v3/calendars/"
  }
}