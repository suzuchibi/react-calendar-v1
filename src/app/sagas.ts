import { put, call, takeEvery } from 'redux-saga/effects';
import { CallEffect, PutEffect } from 'redux-saga/effects';
import { Props as SagaProps } from './sagaActions';
import { Holidays } from '../features/calendar/types';
import GAPI from '../api/googleCalendarApi';

import config from '../config/app.json';
import {
  getYear,
  getMonth,
  getDate,
  startOfMonth,
  endOfMonth,
  add,
} from 'date-fns';

export interface ApiProps {
  type: string;
  payload: SagaProps;
}
export interface SuccessProps {
  type: string;
  payload: {
    mode: string;
    result: Array<Holidays>;
  };
}
export interface ErrorProps {
  type: string;
  payload: {
    mode: string;
  };
}
export type SagaTypes = CallEffect | PutEffect;

/**
 * SetDateText タイムゾーンのテキスト化
 * @param {Date | number} e
 * @return {String}
 */
const setTimeText = (e: Date | number) => {
  const pad = (t: number) => ('00' + t).slice(-2);
  return `${getYear(e)}-${pad(getMonth(e) + 1)}-${pad(getDate(e))}T00:00:00Z`;
};

/**
 * GetTimeZone 開始と終了の日付取得
 * @param {number} stamp
 * @param {String} mode
 * @return {Object<{start: String, end: String}>}
 */
const getTimeZone = (stamp: number, mode: String) => {
  switch (mode) {
    case 'cal':
      const s = startOfMonth(stamp);
      const e = endOfMonth(stamp);
      return {
        start: setTimeText(s),
        end: setTimeText(e),
      };
    case 'schedule':
      const dateAddSix = add(stamp, { days: 7 });
      return {
        start: setTimeText(stamp),
        end: setTimeText(dateAddSix),
      };
    case 'all':
      const sAll = startOfMonth(stamp);
      const end = endOfMonth(stamp);
      const endAddSix = add(end, { days: 6 });
      return {
        start: setTimeText(sAll),
        end: setTimeText(endAddSix),
      };
    default:
      return {
        start: '',
        end: '',
      };
  }
};

// ここからジェネレーター

/**
 * Success 成功した後のディスパッチ処理
 * @param {object} payload
 * @returns {void}
 */
export function* success({
  payload,
}: SuccessProps): Generator<PutEffect, void, SuccessProps> {
  const { mode, result } = payload;
  switch (mode) {
    case 'cal':
      yield put({ type: 'calendar/changeHolidays', payload: result });
      yield put({ type: 'calendar/changeStatus', payload: 'loaded' });
      break;
    case 'schedule':
      yield put({ type: 'schedule/changeHolidays', payload: result });
      yield put({ type: 'schedule/changeStatus', payload: 'loaded' });
      break;
    case 'all':
      yield put({ type: 'calendar/changeHolidays', payload: result });
      yield put({ type: 'schedule/changeHolidays', payload: result });
      yield put({ type: 'calendar/changeStatus', payload: 'loaded' });
      yield put({ type: 'schedule/changeStatus', payload: 'loaded' });
      break;
  }
}

/**
 * Error 失敗した後のディスパッチ処理
 * @param {object} payload modeの値が返る
 * @returns {void}
 */
export function* error({
  payload,
}: ErrorProps): Generator<PutEffect, void, string> {
  const { mode } = payload;
  switch (mode) {
    case 'cal':
      yield put({ type: 'calendar/changeHolidays', payload: [] });
      yield put({ type: 'calendar/changeStatus', payload: 'loaded' });
      break;
    case 'schedule':
      yield put({ type: 'schedule/changeHolidays', payload: [] });
      yield put({ type: 'schedule/changeStatus', payload: 'loaded' });
      break;
    case 'all':
      yield put({ type: 'calendar/changeHolidays', payload: [] });
      yield put({ type: 'schedule/changeHolidays', payload: [] });
      yield put({ type: 'calendar/changeStatus', payload: 'loaded' });
      yield put({ type: 'schedule/changeStatus', payload: 'loaded' });
      break;
  }
}

/**
 * Generator GoogleCalendarApi
 * @param {object} payload
 * @returns {void}
 */
export function* googleCalendarApiSaga({
  payload,
}: ApiProps): Generator<SagaTypes, void, any> {
  const { mode, stamp } = payload;
  const { start, end } = getTimeZone(stamp, mode);
  const param = `timeMin=${start}&timeMax=${end}&singleEvents=true`;
  const encode = encodeURIComponent(config.gapiCal.address);
  const path = `${config.gapiCal.gurl}${encode}/events?${param}`;
  try {
    // GoogleCalendarAPIコール
    const result = yield call(GAPI.start, path);
    yield put({
      type: 'app/googleCalendarApiSuccess',
      payload: { mode: mode, result: result },
    });
  } catch (err) {
    yield put({ type: 'snac/changeMsg', payload: err });
    yield put({ type: 'snac/changeOpen', payload: true });
    yield put({
      type: 'app/googleCalendarApiError',
      payload: { mode: mode },
    });
  }
}

export default function* rootSaga() {
  yield takeEvery('app/googleCalendarAPI', googleCalendarApiSaga);
  yield takeEvery('app/googleCalendarApiSuccess', success);
  yield takeEvery('app/googleCalendarApiError', error);
}
