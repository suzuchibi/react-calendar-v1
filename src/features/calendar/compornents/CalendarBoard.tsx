import React from 'react';
import classNames from 'classnames';
import { withRouter, RouteComponentProps } from 'react-router-dom';
import { connectProps } from '../containers/CalendarBoard';
import { Theme } from '@material-ui/core/styles/createMuiTheme';
import { WithStyles, StyleRules, createStyles } from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';
import blue from '@material-ui/core/colors/blue';
import red from '@material-ui/core/colors/red';
import createCalendar from '../utils/createCalendar';

const useStyles = (theme: Theme): StyleRules =>
  createStyles({
    calBox: {
      width: '100%',
      display: 'flex',
      flexWrap: 'wrap',
    },
    calWeek: {
      width: '100%',
      display: 'flex',
      listStyleType: 'none',
      margin: 0,
      padding: 0,
    },
    calWeekList: {
      width: 'calc(100% / 7)',
      textAlign: 'center',
      lineHeight: '2.0em',
    },
    calLink: {
      color: 'rgba(0, 0, 0, 0.87)',
      textDecoration: 'none',
      borderRadius: 50,
      display: 'block',
      transition: theme.transitions.create(['background'], {
        easing: theme.transitions.easing.sharp,
        duration: theme.transitions.duration.leavingScreen,
      }),
      '&:hover': {
        cursor: 'pointer',
        background: 'rgba(0, 0, 0, 0.1)',
      },
    },
    today: {
      fontWeight: 'bold',
    },
    selected: {
      background: 'rgba(0, 0, 0, 0.1)',
    },
    saturday: {
      color: blue[700],
    },
    sunday: {
      color: red[500],
    },
    holiday: {
      color: red[500],
    },
    glayText: {
      color: '#aaa',
    },
  });

type GetProps = { md?: boolean };
type Props = WithStyles<typeof useStyles> &
  connectProps &
  RouteComponentProps &
  GetProps;

class CalendarBoard extends React.Component<Props> {
  private click(target: string, stamp: number | undefined, md?: boolean) {
    if (stamp) {
      this.props.changeSchedule('pending');
      this.props.asyncGoogleApiSaga('schedule', stamp);
      if (md) this.props.handleDrawer(true);
      if (this.props.anchor !== null) this.props.changeAnchor(null);
      this.props.history.push(target);
    }
  }
  render() {
    const { classes, current, today, holidays, selected, md } = this.props;
    const calendar = createCalendar(current, holidays, today, selected);
    return (
      <div className={classes.calBox}>
        {calendar.map((value, index) => {
          return (
            <ul className={classes.calWeek} key={index}>
              {value.map(
                (
                  v: {
                    type?: String;
                    value?: Number;
                    link?: String;
                    jHoliday?: Boolean;
                    isToday?: Boolean;
                    isSelected?: Boolean;
                    timeStamp?: number;
                  },
                  i
                ) => {
                  if (v.type === 'gray') {
                    return (
                      <li
                        className={classNames([
                          classes.calWeekList,
                          classes.glayText,
                        ])}
                        key={i}
                      >
                        {v.value}
                      </li>
                    );
                  } else {
                    return (
                      <li
                        onClick={() =>
                          this.click(`/?date=${v.link}`, v.timeStamp, md)
                        }
                        key={i}
                        className={classNames(
                          [classes.calWeekList],
                          [classes.calLink],
                          {
                            [classes.sunday]: v.type === 'sunday',
                            [classes.saturday]: v.type === 'saturday',
                            [classes.holiday]: v.jHoliday,
                            [classes.today]: v.isToday,
                            [classes.selected]: v.isSelected,
                          }
                        )}
                      >
                        {v.value}
                      </li>
                    );
                  }
                }
              )}
            </ul>
          );
        })}
      </div>
    );
  }
}

export default withRouter(withStyles(useStyles)(CalendarBoard));
