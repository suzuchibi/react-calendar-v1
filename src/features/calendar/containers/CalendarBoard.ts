import { ThunkDispatch } from 'redux-thunk';
import { connect, ConnectedProps } from 'react-redux';
import CalendarBoard from '../compornents/CalendarBoard';
import { RootState } from '../../../app/reducer';

// Actions
import { changeStatus as changeSchedule } from '../../scheduler/actions/scheduleActions';
import { asyncGoogleCalendarAction } from '../actions/calendarActions';
import { ScheduleActionTypes } from '../../scheduler/types';
import { handleAnchorEl as changeAnchor } from '../../dialog/calendarPoper/actions';
import { PoperActionTypes } from '../../dialog/calendarPoper/types';
import { handleDrawer } from '../../side/isDrawerReducer';
import { DrwerActionTypes } from '../../side/types';

// Combine Types
type ActionTypes = ScheduleActionTypes | PoperActionTypes | DrwerActionTypes;

const mapStateToProps = (state: RootState) => {
  return {
    today: state.today,
    current: state.cal.current,
    holidays: state.cal.holidays,
    selected: state.schedule.selected,
    anchor: state.calPoper.anchorEl,
  };
};

const mapDispatchToProps = (
  dispatch: ThunkDispatch<RootState, undefined, ActionTypes>
) => {
  return {
    changeSchedule(event: String) {
      dispatch(changeSchedule(event));
    },
    changeAnchor(event: HTMLButtonElement | null) {
      dispatch(changeAnchor(event));
    },
    handleDrawer(flag: Boolean) {
      dispatch(handleDrawer(flag));
    },
    asyncGoogleCalendar(stamp: number, mode: String) {
      dispatch(asyncGoogleCalendarAction(stamp, mode));
    },
  };
};

const connecter = connect(mapStateToProps, mapDispatchToProps);
export type connectProps = ConnectedProps<typeof connecter>;
export default connect(mapStateToProps, mapDispatchToProps)(CalendarBoard);
