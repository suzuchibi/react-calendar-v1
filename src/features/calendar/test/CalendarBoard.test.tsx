import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import CalendarBoard from '../containers/CalendarBoard';

import { mockStore as store } from '../../../app/mockStore';

// Termial
// yarn test src/features/calendar/test/CalendarBoard.test.ts

describe('<CalendarBoard>', () => {
  it('スナップショット テスト', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <MemoryRouter>
          <CalendarBoard />
        </MemoryRouter>
      </Provider>
    );
    // snapshot
    expect(asFragment()).toMatchSnapshot();
  });
});
