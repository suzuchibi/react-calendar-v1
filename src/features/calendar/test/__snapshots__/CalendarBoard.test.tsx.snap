// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CalendarBoard> スナップショット テスト 1`] = `
<DocumentFragment>
  <div
    class="CalendarBoard-calBox-1"
  >
    <ul
      class="CalendarBoard-calWeek-2"
    >
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-glayText-10"
      >
        27
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-glayText-10"
      >
        28
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-glayText-10"
      >
        29
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-glayText-10"
      >
        30
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4"
      >
        1
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4"
      >
        2
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4 CalendarBoard-saturday-7"
      >
        3
      </li>
    </ul>
    <ul
      class="CalendarBoard-calWeek-2"
    >
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4 CalendarBoard-sunday-8"
      >
        4
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4"
      >
        5
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4"
      >
        6
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4"
      >
        7
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4"
      >
        8
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4"
      >
        9
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4 CalendarBoard-saturday-7"
      >
        10
      </li>
    </ul>
    <ul
      class="CalendarBoard-calWeek-2"
    >
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4 CalendarBoard-sunday-8"
      >
        11
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4 CalendarBoard-today-5 CalendarBoard-selected-6"
      >
        12
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4"
      >
        13
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4"
      >
        14
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4"
      >
        15
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4"
      >
        16
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4 CalendarBoard-saturday-7"
      >
        17
      </li>
    </ul>
    <ul
      class="CalendarBoard-calWeek-2"
    >
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4 CalendarBoard-sunday-8"
      >
        18
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4"
      >
        19
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4"
      >
        20
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4"
      >
        21
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4 CalendarBoard-holiday-9"
      >
        22
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4"
      >
        23
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4 CalendarBoard-saturday-7"
      >
        24
      </li>
    </ul>
    <ul
      class="CalendarBoard-calWeek-2"
    >
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4 CalendarBoard-sunday-8"
      >
        25
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4"
      >
        26
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4"
      >
        27
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4"
      >
        28
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4"
      >
        29
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4"
      >
        30
      </li>
      <li
        class="CalendarBoard-calWeekList-3 CalendarBoard-calLink-4 CalendarBoard-saturday-7"
      >
        31
      </li>
    </ul>
    <ul
      class="CalendarBoard-calWeek-2"
    />
  </div>
</DocumentFragment>
`;
